### PR TITLE
feat: semver github action

### PR DESCRIPTION
**Pull request by MikhailKorol-saritasa**
Implement `semver` and `process_dotnet` scripts as action
Prepare test workflow on all available runners (github win/linux, self-hosted linux/mac)

Task: https://saritasa.atlassian.net/browse/SD-373

Replacement for https://github.com/saritasa-nest/saritasa-github-actions-getversion
Full test with `process_dotnet` execution can be found here: https://github.com/saritasa-nest/core-source-backend/actions/runs/6312408778/job/17142032808

### CODE REVIEW

File: .github/actions/semver/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+- main branch releases can be versioned only from current commit tag
+- detached HEAD should have tag that belongs main branch
+- release/* releases can be versioned only from branch name but not tag, like release/1.1.1
+- hotfix releases based on main tag (or failover) and postfix as title for semver
+- all others in priority: branch tag -> main tag -> failover.
+That grants us option when we can build well versioned feature/* releases with any tag inside. 
+
+Image below is editable via draw.io
+
+![image](semver.png)
```
Comment (luciano-buono) [inline .github/actions/semver/README.md:17]: Looks like the image isn't uploaded

**Review by luciano-buono — APPROVED** [review]

File: .github/actions/semver/bin/semver [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+#######################################
+# Get commit count (weight) for current branch from first parent
+# Globals:
+#	None
+# Arguments:
+#	None
+# Output:
+#	Commit weight as natural number
+_get_commit_weigh(){
```
Comment (populov) [inline .github/actions/semver/bin/semver:115]: Also need update usages
```suggestion
_get_commit_weight(){
```

**Review by populov — APPROVED** [review]

**Review by udaltsovra — APPROVED** [review]

File: .github/actions/semver/README.md (hunk 1 above)
Comment (kseniyashaydurova) [inline .github/actions/semver/README.md:17]: Shouldn't image.png be at the same level as README.md file?

File: .github/actions/semver/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# Purpose
+
+This is github action with bash based script that provides semantic version generation with specific logic (different than gitversion).
```
Comment (kseniyashaydurova) [inline .github/actions/semver/README.md:3]: Please also add link to the same script we use to build docker image with semver - https://github.com/saritasa-nest/saritasa-devops-docker-images/tree/main/public/semver

That if there would be any changes in `process_dotet` and `semver` logic, `saritasa-devops-docker-images/public/semver` should be also updated

File: .github/actions/semver/action.yaml [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      run: |
+        echo "VERSIONING_DEBUG=true" >> "$GITHUB_ENV"
+    - id: version
+      shell: bash
+      run: |
+        set -o pipefail
+        JSON=$(semver)
+        if [[ -z ${JSON} ]]; then
+          exit -1
+        else
```
Comment (kseniyashaydurova) [inline .github/actions/semver/action.yaml:35]: No need in `else` block here, cause you already did `exit -1` in if. so you can do just like this:

```
set -o pipefail
JSON=$(semver)
if [[ -z ${JSON} ]]; then
  exit -1
fi
echo "json=${JSON}" >> $GITHUB_OUTPUT
```

File: .github/actions/semver/action.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            exit -1
+        fi
+        semver | process_dotnet
+    - id: file
+      shell: bash
+      if: inputs.file != ''
+      run: |
+        echo "${{steps.version.outputs.json}}" >> ${{inputs.file}}
+
+
```
Comment (kseniyashaydurova) [inline .github/actions/semver/action.yaml:64]: Please use linters

File: .github/actions/semver/action.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
+name: 'semver'
+description: 'Provides sevmer output for current git repo'
+inputs:
+  debug:
```
Comment (kseniyashaydurova) [inline .github/actions/semver/action.yaml:4]: Please add description for each input

File: .github/actions/semver/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+# Usage
+- `actions/checkout` with `fetch-depth: 0` is required for proper semver script work
+
+
+```yaml
+      - id: semver
+        uses: saritasa-nest/saritasa-github-actions/.github/actions/semver@action/semver
+      - run: |
+          echo "Building ${{fromJson(steps.semver.outputs.version).SemVer}}"
+      - name: Release app
```
Comment (kseniyashaydurova) [inline .github/actions/semver/README.md:28]: Please add comment here
```suggestion
      # Use semver output in `fastlane-ios` build example
      - name: Release app
```

File: .github/actions/semver/README.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+Image below is editable via draw.io
+
+![image](semver.png)
+
+# Usage
+- `actions/checkout` with `fetch-depth: 0` is required for proper semver script work
+
+
+```yaml
```
Comment (kseniyashaydurova) [inline .github/actions/semver/README.md:23]: Please add more examples with different params that action accepts and example for `dotnet`

File: .github/workflows/test-semver.yaml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - id: semver
+        uses: ./.github/actions/semver
+        with:
+          process_dotnet: true
+      - run: |
+          echo "Computed version: ${{fromJson(steps.semver.outputs.version).InformationalVersion}}"
```
Comment (kseniyashaydurova) [inline .github/workflows/test-semver.yaml:21]: Linters pls

**Review by kseniyashaydurova — CHANGES_REQUESTED** [review]

File: .github/actions/semver/README.md (hunk 1 above)
Comment (MikhailKorol-saritasa) [reply .github/actions/semver/README.md:17]: Fixed

File: .github/actions/semver/README.md (hunk 3 above)
Comment (MikhailKorol-saritasa) [reply .github/actions/semver/README.md:3]: Done

File: .github/actions/semver/action.yaml (hunk 4 above)
Comment (MikhailKorol-saritasa) [reply .github/actions/semver/action.yaml:35]: Done

File: .github/actions/semver/action.yaml (hunk 5 above)
Comment (MikhailKorol-saritasa) [reply .github/actions/semver/action.yaml:64]: Done

File: .github/actions/semver/action.yaml (hunk 6 above)
Comment (MikhailKorol-saritasa) [reply .github/actions/semver/action.yaml:4]: Done

File: .github/actions/semver/README.md (hunk 7 above)
Comment (MikhailKorol-saritasa) [reply .github/actions/semver/README.md:28]: Added as subtitle

File: .github/actions/semver/README.md (hunk 8 above)
Comment (MikhailKorol-saritasa) [reply .github/actions/semver/README.md:23]: Done

File: .github/workflows/test-semver.yaml (hunk 9 above)
Comment (MikhailKorol-saritasa) [reply .github/workflows/test-semver.yaml:21]: Done

**Review by SergeyDeminSaritasa — APPROVED** [review]

**Review by kseniyashaydurova — APPROVED** [review]

**Review by KrawczowaKris — APPROVED** [review]

**Review by DanilaKazakevich — APPROVED** [review]

**Review by roman-mightydevops — APPROVED** [review]

**Review by darliiin — APPROVED** [review]

**Pull request merged**